File: src/ipc/prefs.ts
import { invoke } from "@tauri-apps/api/core";

/**
 * Generic key/value preference store. Backed by SQLite on the Rust side
 * (table `app_prefs`). Use this for prefs that need to be visible to BOTH
 * the front and the Rust setup phase (e.g. `claude_runtime` is read before
 * the sidecar spawns). UI-only flags should keep using localStorage.
 */
export function getPref(key: string): Promise<string | null> {
  return invoke<string | null>("get_pref", { key });
}

export function setPref(key: string, value: string): Promise<void> {
  return invoke<void>("set_pref", { key, value });
}

// Known pref keys — keep in sync with `commands::prefs::KEY_*` on the Rust
// side. Keeping them centralised here avoids typos drifting across files.
export const PREF_CLAUDE_RUNTIME = "claude_runtime";
export type ClaudeRuntimePref = "auto" | "native" | "wsl";

File: src-tauri/src/commands/prefs.rs
//! Generic key/value preference store backed by the `app_prefs` table.
//!
//! Used for settings that need to be reachable from BOTH the frontend (UI
//! toggles) and the Rust setup phase (e.g. `claude_runtime` is read before
//! the sidecar is spawned, which happens before the front is even loaded).
//! Anything that's purely a UI-only flag should keep using localStorage —
//! this exists for prefs that cross the JS/Rust boundary.

use rusqlite::{Connection, OptionalExtension};
use tauri::State;

use crate::db::{DbError, DbState};

/// Value passed to the sidecar via `--claude-runtime`. Default = `auto`,
/// which keeps the existing behaviour (native first, no WSL fallback unless
/// the user opts in).
pub const KEY_CLAUDE_RUNTIME: &str = "claude_runtime";

pub fn read_pref(conn: &Connection, key: &str) -> Result<Option<String>, DbError> {
    let v: Option<String> = conn
        .query_row(
            "SELECT value FROM app_prefs WHERE key = ?1",
            [key],
            |r| r.get(0),
        )
        .optional()?;
    Ok(v)
}

pub fn write_pref(conn: &Connection, key: &str, value: &str) -> Result<(), DbError> {
    conn.execute(
        "INSERT INTO app_prefs(key, value) VALUES(?1, ?2)
         ON CONFLICT(key) DO UPDATE SET value = excluded.value",
        [key, value],
    )?;
    Ok(())
}

#[tauri::command]
pub fn get_pref(state: State<DbState>, key: String) -> Result<Option<String>, String> {
    let conn = state.conn.lock().map_err(|e| e.to_string())?;
    read_pref(&conn, &key).map_err(|e| e.to_string())
}

#[tauri::command]
pub fn set_pref(state: State<DbState>, key: String, value: String) -> Result<(), String> {
    let conn = state.conn.lock().map_err(|e| e.to_string())?;
    write_pref(&conn, &key, &value).map_err(|e| e.to_string())
}
